x/gov/keeper.go

```diff
@@ -0,0 +1,415 @@
+		newDeposit := Deposit{depositer, depositAmount}
+		keeper.setDeposit(ctx, proposalID, depositer, newDeposit)
+	} else {
+		currDeposit.Amount = currDeposit.Amount.Plus(depositAmount)
+		keeper.setDeposit(ctx, proposalID, depositer, *currDeposit)
+	}
+
+	return nil
+}
+
+// Gets the vote of a specific voter on a specific proposal
+func (keeper Keeper) GetDeposits(ctx sdk.Context, proposalID int64) sdk.Iterator {
+	store := ctx.KVStore(keeper.storeKey)
+	return sdk.KVStorePrefixIterator(store, []byte(fmt.Sprintf("%d", proposalID)+":deposits:"))
+}
+
+// Gets the vote of a specific voter on a specific proposal
+func (keeper Keeper) RefundDeposits(ctx sdk.Context, proposalID int64) {
+	store := ctx.KVStore(keeper.storeKey)
+	depositsIterator := keeper.GetDeposits(ctx, proposalID)
+
+	for ; depositsIterator.Valid(); depositsIterator.Next() {
+		deposit := &Deposit{}
+		keeper.cdc.MustUnmarshalBinary(depositsIterator.Value(), deposit)
+
+		_, _, err := keeper.ck.AddCoins(ctx, deposit.Depositer, deposit.Amount)
+		if err != nil {
+			panic("should not happen")
+		}
+
+		store.Delete(depositsIterator.Key())
+	}
+
+	depositsIterator.Close()
+}
+
+// Gets the vote of a specific voter on a specific proposal
+func (keeper Keeper) DeleteDeposits(ctx sdk.Context, proposalID int64) {
+	store := ctx.KVStore(keeper.storeKey)
+	depositsIterator := keeper.GetDeposits(ctx, proposalID)
+
+	for ; depositsIterator.Valid(); depositsIterator.Next() {
+		store.Delete(depositsIterator.Key())
+	}
+
+	depositsIterator.Close()
+}
+
+// =====================================================
+// ProposalQueues
+
+func (keeper Keeper) getActiveProposalQueue(ctx sdk.Context) ProposalQueue {
+	store := ctx.KVStore(keeper.storeKey)
+	bz := store.Get([]byte("activeProposalQueue"))
+	if bz == nil {
+		return nil
+	}
+
+	proposalQueue := &ProposalQueue{}
+	err := keeper.cdc.UnmarshalBinary(bz, proposalQueue) // TODO: switch to UnmarshalBinaryBare when new go-amino gets added
+	if err != nil {
+		panic(err)
+	}
+
+	return *proposalQueue
+}
+
+func (keeper Keeper) setActiveProposalQueue(ctx sdk.Context, proposalQueue ProposalQueue) {
+	store := ctx.KVStore(keeper.storeKey)
+
+	bz, err := keeper.cdc.MarshalBinary(proposalQueue) // TODO: switch to MarshalBinaryBare when new go-amino gets added
+	if err != nil {
+		panic(err)
+	}
+
+	store.Set([]byte("activeProposalQueue"), bz)
+}
+
+// Return the Proposal at the front of the ProposalQueue
+func (keeper Keeper) ActiveProposalQueuePeek(ctx sdk.Context) *Proposal {
+	proposalQueue := keeper.getActiveProposalQueue(ctx)
+	if len(proposalQueue) == 0 {
+		return nil
+	}
+	return keeper.GetProposal(ctx, proposalQueue[0])
+}
+
+// Remove and return a Proposal from the front of the ProposalQueue
+func (keeper Keeper) ActiveProposalQueuePop(ctx sdk.Context) *Proposal {
+	proposalQueue := keeper.getActiveProposalQueue(ctx)
+	if len(proposalQueue) == 0 {
+		return nil
+	}
+	frontElement, proposalQueue := proposalQueue[0], proposalQueue[1:]
```

Comment:
This looks like it will panic if `len(proposalQueue) == 1` should update logic to prevent this

Reply:
It doesn't.  Go allows you to do this.  See https://play.golang.org/p/2rVYOmh9XyQ
Also, this is already checked in the tests.

Reply:
cool